### PR TITLE
Modify setchplenv.* - can be sourced from anywhere

### PR DESCRIPTION
**Pull request by ben-albrecht**
* The `setchplenv.{bash, csh, fish}` scripts can now be sourced from anywhere.
* `util/config/fixpath` perl script rewritten as a Python script: `util/config/fixpath.py`
    * `$CHPL_HOME` grepped to confirm nothing else depends on old `fixpath`
* `setchplenv.sh` remains for POSIX-standard shells (such as dash), and still requires to be sourced from `$CHPL_HOME` (no known workaround for this at the moment)
* All changes tested in bash, csh, and fish, sh (OS X's sh-compatible bash), and dash

### CODE REVIEW

**Comment by ben-albrecht** [comment]
@Kyle-B

File: util/setchplenv.fish [hunk 1]
```diff
@@ -1,34 +1,38 @@
...
-echo "to $CHPL_HOME"
-
 echo -n "Setting CHPL_HOST_PLATFORM "
-set -x CHPL_HOST_PLATFORM (eval $CHPL_HOME/util/chplenv/chpl_platform.py)
+set -x CHPL_HOST_PLATFORM (eval "$CHPL_HOME/util/chplenv/chpl_platform.py")
 echo "to $CHPL_HOST_PLATFORM"
 
 echo -n "Updating PATH "
-set -x PATH "$CHPL_HOME/bin/$CHPL_HOST_PLATFORM" "$CHPL_HOME/util" $MYPATH
+eval set -x PATH $MYPATH $CHPL_HOME/bin/$CHPL_HOST_PLATFORM $CHPL_HOME/util
```
Comment (Kyle-B) [inline util/setchplenv.fish]: Why the order flip here? The `eval` also seems fishy.

File: util/config/fixpath.py [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+
+    if len(sys.argv) < 2:
+        # Return empty string if len(path) == 0
+        newpath = ""
+    else:
+        path = sys.argv[1]
+        if ":" in path:
+            # bash/csh/tcsh-like path
+            newpath = main(path, sep=":")
+        elif " " in path:
```
Comment (Kyle-B) [inline util/config/fixpath.py]: This isn't safe, and is going to break if you have a space in directory name that is in your path.

File: util/setchplenv.fish (hunk 1 above)
Comment (ben-albrecht) [inline util/setchplenv.fish]: * order flip - Unintentional result of some trial and error testing 

* eval - necessary in fish from what I understand.

File: util/config/fixpath.py (hunk 2 above)
Comment (ben-albrecht) [inline util/config/fixpath.py]: Good point - no need to make the script user friendly, if it's only being called from setchplenv anyway. This is resolved in latest commit.

File: util/config/fixpath.py [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+    """
+    Removes path components that begin with $CHPL_HOME, to reduce
+    $PATH & $MANPATH pollution
+
+    :path: $PATH or $MANPATH
+    :sep: Expected path separator
+    :returns: new path with $CHPL_HOME components stripped
+    """
+    chpl_home = os.getenv("CHPL_HOME")
+    newpath = [p for p in path.split(sep) if chpl_home not in p]
```
Comment (Kyle-B) [inline util/config/fixpath.py]: The other issue with spaces in the path shows up here. If my path looks like: `/bin /usr/bin /home/Kyle\ B/$CHPL_HOME` PATH is going to get corrupted.

File: util/config/fixpath.py [hunk 4]
```diff
@@ -0,0 +1,68 @@
+#!/usr/bin/env python
+
+from __future__ import print_function
+
+import os
+from sys import argv
+
+
+def splitescaped(s, delim=":", escape='\\'):
```
Comment (Kyle-B) [inline util/config/fixpath.py]: I probably would do with with something like:

```python
pattern = r'(?<!{0}){1}'.format(escape, delim)
re.split(pattern, s)
```
this wont support escaping your escape character (same as your way, which is probably fine), but I find it much easier to read.

File: util/quickstart/setchplenv.fish [hunk 5]
```diff
@@ -1,36 +1,40 @@
...
-echo "to $CHPL_HOME"
-
 echo -n "Setting CHPL_HOST_PLATFORM "
-set -x CHPL_HOST_PLATFORM (eval $CHPL_HOME/util/chplenv/chpl_platform.py)
+set -x CHPL_HOST_PLATFORM (eval "$CHPL_HOME/util/chplenv/chpl_platform.py")
 echo "to $CHPL_HOST_PLATFORM"
 
 echo -n "Updating PATH "
-set -x PATH "$CHPL_HOME/bin/$CHPL_HOST_PLATFORM" "$CHPL_HOME/util" $MYPATH
+eval set -x PATH $CHPL_HOME/bin/$CHPL_HOST_PLATFORM $CHPL_HOME/util $MYPATH
```
Comment (Kyle-B) [inline util/quickstart/setchplenv.fish]: I'd prefer the `eval`s be moved up to `set MYPATH` and `set MYMANPATH` lines, it is clearer why they are required when placed there.

File: util/setchplenv.fish [hunk 6]
```diff
@@ -1,34 +1,38 @@
...
-echo "to $CHPL_HOME"
-
 echo -n "Setting CHPL_HOST_PLATFORM "
-set -x CHPL_HOST_PLATFORM (eval $CHPL_HOME/util/chplenv/chpl_platform.py)
+set -x CHPL_HOST_PLATFORM (eval "$CHPL_HOME/util/chplenv/chpl_platform.py")
 echo "to $CHPL_HOST_PLATFORM"
 
 echo -n "Updating PATH "
-set -x PATH "$CHPL_HOME/bin/$CHPL_HOST_PLATFORM" "$CHPL_HOME/util" $MYPATH
+eval set -x PATH $CHPL_HOME/bin/$CHPL_HOST_PLATFORM $CHPL_HOME/util $MYPATH
```
Comment (Kyle-B) [inline util/setchplenv.fish]: Same as in quickstart version.

File: util/quickstart/setchplenv.fish (hunk 5 above)
Comment (ben-albrecht) [inline util/quickstart/setchplenv.fish]: Fixed in latest commit

File: util/config/fixpath.py (hunk 4 above)
Comment (ben-albrecht) [inline util/config/fixpath.py]: Just when I thought I had a clever solution....  I switched to this approach in latest commit.

**Comment by Kyle-B** [comment]
Looks good.

**Pull request merged**